### PR TITLE
This PR solve the issue #505 and #512

### PR DESCRIPTION
**Pull request by MathiasPaulin**
There is a PR on Radium Apps that follow this one.

UPDATE the form below to describe your PR.

* **Please check if the PR fulfills these requirements**
- [ ] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

Be aware that the PR request cannot be accepted if it doesn't pass the Continuous Integration tests.

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
Code refactoring to have a better usage of the Material interface for gizmos rendering.
Fix link problem on shaders for undefined fragment input

* **What is the current behavior?** (You can also link to an open issue here)

* **What is the new behavior (if this is a feature change)?**

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)

* **Other information**:

### CODE REVIEW

File: src/GuiBase/Viewer/Gizmo/Gizmo.cpp [hunk 1]
```diff
@@ -68,20 +85,64 @@ bool Gizmo::findPointOnPlane( const Engine::Camera& cam,
...
+    m_renderParameters.addParameter( "material.color", m_selectedColor );
+}
+
+void Gizmo::UiSelectionControler::clearState() {
+    m_selected = false;
+    // Todo : find a way to remove this dynamic cast
+    auto m = dynamic_cast<Engine::PlainMaterial*>( m_associatedMaterial.get() );
+    CORE_ASSERT( m != nullptr, "UiSelectionControler only designed for Plain Material." );
+    m_renderParameters.addParameter( "material.color", m->m_color );
+}
```
Comment (nmellado) [inline src/GuiBase/Viewer/Gizmo/Gizmo.cpp]: Codacy Issue found: [Either the condition 'm!=nullptr' is redundant or there is possible null pointer dereference: m.](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5264830)

**Comment by nmellado** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Clones added
============
- src/GuiBase/Viewer/Gizmo/RotateGizmo.cpp  1
         

Clones removed
==============
+ src/GuiBase/Viewer/Gizmo/TranslateGizmo.cpp  -4
+ src/GuiBase/Viewer/Gizmo/ScaleGizmo.cpp  -6
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5264830&bid=17247385)

**Comment by nmellado** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Clones added
============
- src/GuiBase/Viewer/Gizmo/RotateGizmo.cpp  1
         

Clones removed
==============
+ src/GuiBase/Viewer/Gizmo/ScaleGizmo.cpp  -6
+ src/GuiBase/Viewer/Gizmo/TranslateGizmo.cpp  -4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5264830&bid=17247385)

File: src/GuiBase/Viewer/Gizmo/Gizmo.cpp [hunk 2]
```diff
@@ -68,20 +82,61 @@ bool Gizmo::findPointOnPlane( const Engine::Camera& cam,
...
+void Gizmo::UiSelectionControler::toggleState() {
+    m_selected = !m_selected;
+    if ( m_selected ) { m_renderParameters.addParameter( "material.color", m_selectedColor ); }
+    else
+    {
+        // Todo : find a way to remove this dynamic cast
+        auto m = dynamic_cast<Engine::PlainMaterial*>( m_associatedMaterial.get() );
+        CORE_ASSERT( m != nullptr, "UiSelectionControler only designed for Plain Material." );
+        m_renderParameters.addParameter( "material.color", m->m_color );
+    }
```
Comment (nmellado) [inline src/GuiBase/Viewer/Gizmo/Gizmo.cpp]: Codacy Issue found: [Either the condition 'm!=nullptr' is redundant or there is possible null pointer dereference: m.](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5264830)

**Comment by nmellado** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/GuiBase/Viewer/Gizmo/RotateGizmo.cpp  1
         

Clones removed
==============
+ src/GuiBase/Viewer/Gizmo/TranslateGizmo.cpp  -4
+ src/GuiBase/Viewer/Gizmo/ScaleGizmo.cpp  -6
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5264830&bid=17247385)

**Comment by nmellado** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/GuiBase/Viewer/Gizmo/RotateGizmo.cpp  1
         

Clones removed
==============
+ src/GuiBase/Viewer/Gizmo/ScaleGizmo.cpp  -6
+ src/GuiBase/Viewer/Gizmo/TranslateGizmo.cpp  -4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5264830&bid=17247385)

**Comment by nmellado** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/GuiBase/Viewer/Gizmo/RotateGizmo.cpp  1
         

Clones removed
==============
+ src/GuiBase/Viewer/Gizmo/TranslateGizmo.cpp  -4
+ src/GuiBase/Viewer/Gizmo/ScaleGizmo.cpp  -6
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5264830&bid=17247385)

File: src/GuiBase/Viewer/Gizmo/ScaleGizmo.cpp [hunk 3]
```diff
@@ -30,40 +30,32 @@ ScaleGizmo::ScaleGizmo( Engine::Component* c,
...
-
     for ( uint i = 0; i < 3; ++i )
     {
-        Core::Utils::Color arrowColor         = Core::Utils::Color::Black();
-        arrowColor[i]                         = 1_ra;
         Core::Vector3 cylinderEnd             = Core::Vector3::Zero();
         cylinderEnd[i]                        = ( 1_ra - arrowFrac );
         Core::Geometry::TriangleMesh cylinder = Core::Geometry::makeCylinder(
-            Core::Vector3::Zero(), arrowScale * cylinderEnd, radius, 32, arrowColor );
+            Core::Vector3::Zero(), arrowScale * cylinderEnd, radius, 32 );
```
Comment (hoshiryu) [inline src/GuiBase/Viewer/Gizmo/ScaleGizmo.cpp]: Also remove the attribName l.16

**Comment by MathiasPaulin** [comment]
> All works fine.
> I still have an issue with the PlainMaterial shaders:
> 
> This comes from the fact `in_tangent` is requested in shader file `VertexAttribInterface.frag.glsl` l.10, which is included in both `Plain.frag.glsl` and `PlainZPrepassfrag.glsl`

I don't understand why you are the only one to have this problem (and only with Plain, the same problem would also occur on other programs).
According to the OpenGL Specification (https://www.khronos.org/registry/OpenGL/specs/gl/glspec41.core.pdf, p. 64, "Shader Interface 
 Matching") there must be no error at link time : 
> With separable program objects, interfaces between shader stages may involve the outputs from one program object and the inputs from a second program object. For such interfaces, it is not possible to detect mismatches at link time, because the programs are linked separately. When each such program is linked, all inputs or outputs interfacing with another program stage are treated as active. The linker will generate an executable that assumes the presence of a compatible program on the other side of the interface. If a mismatch between programs occurs, no GL error will be generated, but some or all of the inputs on the interface will be undefined.

Post an issue on that giving  the exact OpenGL version and driver you have. As soon as the issue will be posted, I'll propose a fix. But you will be the only one to be able to test.

File: src/GuiBase/Viewer/Gizmo/Gizmo.cpp [hunk 4]
```diff
@@ -68,20 +82,49 @@ bool Gizmo::findPointOnPlane( const Engine::Camera& cam,
...
+}
+
+Gizmo::UiSelectionControler::UiSelectionControler( std::shared_ptr<Ra::Engine::Material>& material,
+                                                   const Core::Utils::Color& selectedColor ) :
+    ShaderParameterProvider(), m_associatedMaterial( material ), m_selectedColor( selectedColor ) {}
+
+void Gizmo::UiSelectionControler::updateGL() {
+    m_associatedMaterial->updateGL();
+    m_renderParameters = m_associatedMaterial->getParameters();
+    if ( m_selected ) { m_renderParameters.addParameter( "material.color", m_selectedColor ); }
```
[the remark this answers is not in the document]
Comment (MathiasPaulin) [reply src/GuiBase/Viewer/Gizmo/Gizmo.cpp]: To keep some ability to use another material that define a "color" attrib. But, for now, OK, let's reduce the Material to PlainMaterial.

**Comment by nmellado** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/GuiBase/Viewer/Gizmo/RotateGizmo.cpp  1
         

Clones removed
==============
+ src/GuiBase/Viewer/Gizmo/ScaleGizmo.cpp  -6
+ src/GuiBase/Viewer/Gizmo/TranslateGizmo.cpp  -4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5264830&bid=17247385)

File: src/Engine/Renderer/RenderTechnique/ShaderProgram.cpp [hunk 5]
```diff
@@ -25,6 +25,59 @@ namespace Engine {
 
 using namespace Core::Utils; // log
 
+// The two following methods are indemendant of any ShaderProgram object. Make them standard local
+// function to remove dependency on openGL.h for the class header
```
Comment (nmellado) [inline src/Engine/Renderer/RenderTechnique/ShaderProgram.cpp]:
```suggestion
/// The two following methods are independent of any ShaderProgram object. 
/// \todo Change as local function to remove dependency with openGL.h for the class header
```

**Comment by nmellado** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/GuiBase/Viewer/Gizmo/RotateGizmo.cpp  1
         

Clones removed
==============
+ src/GuiBase/Viewer/Gizmo/TranslateGizmo.cpp  -4
+ src/GuiBase/Viewer/Gizmo/ScaleGizmo.cpp  -6
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5264830&bid=17247385)

File: src/Engine/Renderer/RenderTechnique/ShaderProgram.cpp (hunk 5 above)
Comment (MathiasPaulin) [reply src/Engine/Renderer/RenderTechnique/ShaderProgram.cpp]: Well, this is not a todo. This is a Fixed ;)

**Comment by nmellado** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/GuiBase/Viewer/Gizmo/RotateGizmo.cpp  1
         

Clones removed
==============
+ src/GuiBase/Viewer/Gizmo/ScaleGizmo.cpp  -6
+ src/GuiBase/Viewer/Gizmo/TranslateGizmo.cpp  -4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5264830&bid=17247385)

**Comment by nmellado** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/GuiBase/Viewer/Gizmo/RotateGizmo.cpp  1
         

Clones removed
==============
+ src/GuiBase/Viewer/Gizmo/ScaleGizmo.cpp  -6
+ src/GuiBase/Viewer/Gizmo/TranslateGizmo.cpp  -4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5264830&bid=17247385)

**Comment by nmellado** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/GuiBase/Viewer/Gizmo/RotateGizmo.cpp  1
         

Clones removed
==============
+ src/GuiBase/Viewer/Gizmo/ScaleGizmo.cpp  -6
+ src/GuiBase/Viewer/Gizmo/TranslateGizmo.cpp  -4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5264830&bid=17247385)

**Comment by MathiasPaulin** [comment]
@hoshiryu Is it OK with the shader or is there always a problem (with in_viewVector or in_lightVector) ?

**Comment by hoshiryu** [comment]
It works well

**Comment by MathiasPaulin** [comment]
I've a last commit to push on this PR and all will be OK.

**Comment by nmellado** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/GuiBase/Viewer/Gizmo/RotateGizmo.cpp  1
         

Clones removed
==============
+ src/GuiBase/Viewer/Gizmo/ScaleGizmo.cpp  -6
+ src/GuiBase/Viewer/Gizmo/TranslateGizmo.cpp  -4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/STORM/Radium-Engine/pullRequest?prid=5264830&bid=17247385)

**Pull request merged**